File: src/Functional/Haskell/ChainOfResponsibility.hs
data Handler = Handler { name :: String, accepts :: Int -> Bool }

handlers :: [Handler]
handlers =
  [ Handler "faq" (<= 50)
  , Handler "billing" (<= 500)
  , Handler "escalation" (const True)
  ]

handle :: Int -> [Handler] -> [String] -> ([String], String, String)
handle _ [] _ = error "unhandled request"
handle amount (h:rest) visited =
  let visited' = visited ++ [name h]
  in if accepts h amount
       then (visited', name h, "refund(" ++ show amount ++ ")")
       else handle amount rest visited'

main :: IO ()
main = do
  let (visited, handled, result) = handle 250 handlers []
  putStrLn $ "visited=" ++ joinWith ">" visited ++ ";handled=" ++ handled ++ ";result=" ++ result

joinWith :: String -> [String] -> String
joinWith _ [] = ""
joinWith _ [x] = x
joinWith sep (x:xs) = x ++ sep ++ joinWith sep xs

File: src/Functional/Scala/ChainOfResponsibility.scala
final case class Handler(
    name: String,
    accepts: Int => Boolean,
    next: Option[Handler]
) {
  def handle(amount: Int, visited: List[String] = Nil): (List[String], String) = {
    val visitedNow = visited :+ name
    if (accepts(amount)) {
      (visitedNow, name)
    } else {
      next match {
        case Some(handler) => handler.handle(amount, visitedNow)
        case None => throw new IllegalStateException("No handler accepted the request")
      }
    }
  }
}

object ChainOfResponsibilityExample extends App {
  val escalation = Handler("escalation", _ => true, None)
  val billing = Handler("billing", _ <= 500, Some(escalation))
  val faq = Handler("faq", _ <= 50, Some(billing))

  val amount = 250
  val (visited, handled) = faq.handle(amount)
  println(
    "visited=" + visited.mkString(">") +
      ";handled=" + handled +
      ";result=refund(" + amount + ")"
  )
}
